### ISSUE TITLE
Missing messages on unsleeping laptop until refresh

### ISSUE THREAD

**Issue by Kegsay** [vector-im/riot-web#3375]
From @ara4n:

> all messages between $14886737321221486bgccE:matrix.org and $14887082401363892ElEzc:matrix.org in !YnAbbwWklMaoABEKWk:matrix.org were missing, until I hit refresh, at which point they reappeared in the timeline.

Relevant logs from directory `2017-03-05/165932`:
```
2017-03-05T00:29:43.044Z I Completed sync, next_batch=s107046335_163739780_359278_15670637_158574_56324_517545_258185
2017-03-05T00:29:43.045Z I Starting sync since=s107046335_163739780_359278_15670637_158574_56324_517545_258185
2017-03-05T00:29:43.482Z I Completed sync, next_batch=s107046338_163739798_359279_15670638_158574_56324_517545_258185
2017-03-05T00:29:43.483Z I Starting sync since=s107046338_163739798_359279_15670638_158574_56324_517545_258185
2017-03-05T00:29:43.626Z E /sync error %s Error: CORS request rejected: 
2017-03-05T00:29:43.626Z E Error: CORS request rejected: 
2017-03-05T00:29:43.626Z I Starting keep-alive
2017-03-05T00:29:43.627Z I MatrixClient sync state => %s RECONNECTING
2017-03-05T00:44:48.246Z I MatrixClient sync state => %s ERROR
2017-03-05T00:44:48.257Z E Failed to poll for update [object Object]

[ no messages (slept) ]

2017-03-05T16:17:26.262Z I MatrixClient sync state => %s ERROR
2017-03-05T16:17:26.265Z E Failed to poll for update [object Object]
2017-03-05T16:17:26.271Z E Failed to get TURN URIs
2017-03-05T16:17:34.480Z I Starting sync since=s107046338_163739798_359279_15670638_158574_56324_517545_258185
2017-03-05T16:17:44.671Z I Completed sync, next_batch=s107289778_163983507_417860_15710583_159140_56458_519060_261466

```

The same sync token was used later on (`s107046338_163739798_359279_15670638_158574_56324_517545_258185`) and there doesn't seem to be any obvious stack traces in the logs.

**Comment by ara4n** [comment]
it's not clear where on earth the problem is
it's conceivable that synapse went awry and didn't sync the events
or it's conceivable that synapse synced the events fine and then riot-web threw an (invisible?) wobbly and didn't show them in the UI.

@kegsay, @erikjohnston, do you have /any/ hunch on which side of the fence this lies?

**Comment by erikjohnston** [comment]
Did it get a limited sync?

**Comment by Kegsay** [comment]
I've not seen anything like this before. I doubt it is related to Synapse or else more people would've complained about having missed events, plus it wouldn't just magically fix itself on refresh (due to indexeddb caching).

My hunch is that this produced a limited sync which was handled poorly by Riot, for some reason. I have no evidence to support this though.

**Comment by ara4n** [comment]
Just saw this again (ironically in a meeting with @kegsay, @erikjohnston & @dbkr).  My browser crashed overnight (as always), so I refreshed it this morning at around 09:40.  I then tried to read scrollback in Internal, and beyond ~20 messages i had a hole in the timeline which missed a good chunk of messages.  On refreshing it fixed.

It looks a bit as if the wrong prev_batch_token was stored by the indexeddb code, such that it back-paginated me to the wrong point in history when I first loaded the app at 09:40 (presumably an old prev_batch_token), thus breaking pagination?

This is pretty catastrophic as it's the first time our timeline itself has become unreliable and flakey, so bumping priority accordingly.

**Comment by Kegsay** [comment]
I am reasonably certain I have fixed this issue now. The underlying cause was that I failed to implement part of the `/sync` API in the `SyncAccumulator` class.

`/sync` may return `limited: true` if the number of timeline events is greater than the filter `limit` supplied. This means there is a gap in the timeline between the events just received by `/sync` and whatever may be stored in the client. Clients are supposed to purge their stored timeline in this scenario, to avoid having gaps in an otherwise seemingly contiguous timeline.

I just completely forgot to handle this scenario. Now fixed in the JS SDK.
